src/composeinfo.cc

```diff
@@ -0,0 +1,44 @@
+#include "composeinfo.h"
+#include <glib.h>
+#include "logging/logging.h"
+#include "yaml2json.h"
+
+namespace Docker {
+
+ComposeInfo::ComposeInfo(const std::string& yaml) : json_(yaml) {}
+
+bool ComposeInfo::getServices(std::vector<Json::Value>& services) {
+  Json::Value p = json_.root_["services"];
+
+  for (Json::ValueIterator ii = p.begin(); ii != p.end(); ++ii) {
+    services.push_back(ii.key());
+  }
+
+  return !services.empty();
+}
+
+bool ComposeInfo::getServiceImage(const Json::Value& service, std::string& image) {
+  image = json_.root_["services"][service.asString()]["image"].asString();
+  return !image.empty();
+}
+
+bool ComposeInfo::getServiceHash(const Json::Value& service, std::string& hash) {
+  hash = json_.root_["services"][service.asString()]["labels"]["io.compose-spec.config-hash"].asString();
+  return !hash.empty();
+}
+
+int ComposeInfo::getNbrImages() {
```

Comment:
I suspect you did this because Mike's original code was counting "images". The reason he did this was that counting "services" in YAML by straight string parsing isn't really feasible. We have a requirement that every service must have an image, so he can basically grep for images and get the total number of *services*.

tldr; we need a `getNumServices` method which is close to what you have below but can become simpler by dropping the `!image.empty()` stuff

Reply:
+1, there is no need in neither getNbrImages() nor in getNumServices() since there is `getServices(std::vector<Json::Value>& services)` and services.size().

Reply:
yes I agree, it is pointless (and ugly). will remove